Fixes to ArrayLabels documentation

ewanas opened the pull request:
- Removal of string arguments from exceptions in docs
- Changed references to `ArrayLabels` instead of `Array` whenever
  a labeled argument is present
- Fixes to code examples using the wrong argument(wrong name, or missing label)

There are many other references to `Array` in the documentation. Not sure if those should be changed as well.

ewanas committed:
Fixes to ArrayLabels documentation
- Removal of string arguments from exceptions in docs
- Changed references to ArrayLabels instead of Array whenever
  a labeled argument is present
- Fixes to code examples using the wrong argument(wrong name, or missing label)

No change entry needed

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -22,21 +22,21 @@ external length : 'a array -> int = "%array_length"
 
-   Raise [Invalid_argument "index out of bounds"]
-   if [n] is outside the range 0 to [(Array.length a - 1)]. *)
+   Raise [Invalid_argument]
```
```suggestion
   @raise Invalid_argument
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -22,21 +22,21 @@ external length : 'a array -> int = "%array_length"
 
-   Raise [Invalid_argument "index out of bounds"]
-   if [n] is outside the range 0 to [Array.length a - 1]. *)
+   Raise [Invalid_argument]
```
```suggestion
   @raise Invalid_argument
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -79,9 +79,9 @@ val make_matrix : dimx:int -> dimy:int -> 'a -> 'a array array
-  [@@ocaml.deprecated "Use Array.make_matrix instead."]
-(** @deprecated [Array.create_matrix] is an alias for
-   {!Array.make_matrix}. *)
+  [@@ocaml.deprecated "Use ArrayLabels.make_matrix instead."]
```
I think this should stay as `Array.make_matrix`

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -79,9 +79,9 @@ val make_matrix : dimx:int -> dimy:int -> 'a -> 'a array array
-   {!Array.make_matrix}. *)
+  [@@ocaml.deprecated "Use ArrayLabels.make_matrix instead."]
+(** @deprecated [ArrayLabels.create_matrix] is an alias for
+   {!ArrayLabels.make_matrix}. *)
```
```suggestion
   {!make_matrix}. *)
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -91,37 +91,37 @@ val concat : 'a array list -> 'a array
    of array [a].
 
-   Raise [Invalid_argument "Array.sub"] if [start] and [len] do not
+   Raise [Invalid_argument] if [pos] and [len] do not
```
```suggestion
   @raise Invalid_argument if [pos] and [len] do not
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -91,37 +91,37 @@ val concat : 'a array list -> 'a array
+   storing [x] in elements number [pos] to [pos + len - 1].
 
-   Raise [Invalid_argument "Array.fill"] if [ofs] and [len] do not
+   Raise [Invalid_argument] if [pos] and [len] do not
```
```suggestion
   @raise Invalid_argument if [pos] and [len] do not
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -91,37 +91,37 @@ val concat : 'a array list -> 'a array
-   Raise [Invalid_argument "Array.blit"] if [o1] and [len] do not
-   designate a valid subarray of [v1], or if [o2] and [len] do not
-   designate a valid subarray of [v2]. *)
+   Raise [Invalid_argument] if [src_pos] and [len] do not
```
```suggestion
   @raise Invalid_argument if [src_pos] and [len] do not
```

dra27 reviewed (changes requested):
Thanks for taking this on!

The reason this module refers to `Array` and not `ArrayLabels` is because it's intended to be used by `open StdLabels`, where the `Array`, `Bytes`, `List` and `String` modules are all replaced by their labelled counterparts. So the paradigm is:

```ocaml
open StdLabels

let everything = Array.create_matrix ~dimx:42 ~dimy:42 42
```

and hence the deprecation warning more meaningfully refers to `Array.make_matrix` because it's expected that's what you'd write (so the `@@ocaml.deprecated` attributes should definitely keep the `Array.` references).

For the rest (and for the other labelled modules) I'd suggest we could get rid of the module name entirely which removes the ambiguity (and can be applied to the other functions, as per your question) - so instead of `[Array.make_matrix …]` just have `[make_matrix …]` and the same with the cross-references (so instead of `{!ArrayLabels.make_matrix}` just `{!make_matrix}`). What do you think of that, @Octachron?

It's very good to remove the strings from the `Invalid_argument` descriptions, as it discourages matching on them - please could you upgrade them completely to use `@raise`, though (the documentation comments pre-date the availability of `@raise`, I think). I've put those as code suggestions.

Octachron commented:
@dra27, removing the `Array` prefix  sounds like a good idea.

ewanas committed:
Replaced old exception documentation in ArrayLabels with `@raise`
Deprecated warnings should point to `Array` alternatives rather than
those in `ArrayLabels`
Removed module references from function docs to avoid confusing between
StdLabels, Array and ArrayLabels

No change entry needed

ewanas commented:
@dra27 Thank you for checking my PR and for explaining the rationale behind `ArrayLabels` referencing `Array` in the docs.

Do you think `ArrayLabels` should mention something about `StdLabels`? copy-pasting your example is tempting!

dra27 commented:
Thanks for the changes, @foocraft. The remaining thing is to remove all instances of `Array.` and `ArrayLabels.` from the documentation comments - looking at it, `ArrayLabels.` shouldn't appear anywhere and `Array.` should only be present in the `@@deprecated` attribute text.

It would be good to add some text at the top of the module explaining the usage, yes. I would move the `(** Array operations. *)` documentation comment to be before the array type and then expand it with the example - it'd be good if the description included a reference to the `StdLabels` module (i.e. `{!StdLabels}`). As a bonus, if you wanted, you could add that change to the other 3 modules 🙂

foocraft committed:
- Removal of all references to either Array or ArrayLabels from the
  ArrayLabels documentation
- Only references remaining are for the warnings, and they refer to
  Array rather than ArrayLabels

foocraft committed:
Updated documentation for StringLabels, ArrayLabels, ListLabels and BytesLabels to show StdLabels usage

ewanas commented:
Hi @dra27, I've pushed some changes, not exactly sure if it's the right formatting style or not; more on that feedback would be great. For example, what's a good example code to put up there?

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -13,205 +13,214 @@
 
+(** Array operations
+
+   This module is intended to be used through {!StdLabels} which would replace
```
```suggestion
   This module is intended to be used via {!StdLabels} which replaces
```

dra27 commented on stdlib/arrayLabels.mli:
```diff
@@ -13,205 +13,214 @@
+   This module is intended to be used through {!StdLabels} which would replace
+   {!Array}, {!Bytes}, {!List} and {!String} with their labeled counterparts
+
+   As an example usage:
```
```suggestion
   For example:
```

dra27 reviewed (approved):
This looks very nice to me - I've made rewording suggestions for the change to the headers which you *could* apply to the four of them if you wish.

Thanks for your time on this! @Octachron - please merge if you're happy, too

ewanas committed:
Comment fixes

Co-Authored-By: David Allsopp <david.allsopp@metastack.com>

ewanas commented:
I updated across all files. Now it's more coherent :) Thanks for the suggestions.

Octachron commented:
Merged, thanks for the PR!
